packages/flutter/lib/src/services/asset_resolver.dart

```diff
@@ -0,0 +1,85 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+import 'dart:collection';
+
+import 'asset_bundle.dart';
+
+/// Base class for resolver components that transform an input into part of
+/// a resolved asset's key.
+abstract class Qualifier {
+  String resolve();
+}
+
+/// Qualifier that knows how to find the nearest match among a list of
+/// candidates, based on a [SplayTreeMap] of normative values to candidates.
+class BreakpointQualifier extends Qualifier {
+  BreakpointQualifier({ this.candidates, this.value }) {
+    assert(this.candidates.length > 0);
+  }
+  final SplayTreeMap<double, String> candidates;
+  final double value;
+  String resolve() {
+    if (candidates.containsKey(value))
+      return candidates[value];
+    double lower = candidates.lastKeyBefore(value);
+    double upper = candidates.firstKeyAfter(value);
+    if (lower == null)
+      return candidates[upper];
+    if (upper == null)
+      return candidates[lower];
+    if (value > (lower + upper) / 2)
+      return candidates[upper];
+    else
+      return candidates[lower];
+  }
```

Comment:
You definitely want a cache in here (probably for just one value) since you're almost always going to be queried with exactly the same device pixel ratio as last time.